OrbitVulkanLayer/VulkanLayerController.h

```diff
@@ -0,0 +1,460 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_VULKAN_LAYER_VULKAN_LAYER_CONTROLLER_H_
+#define ORBIT_VULKAN_LAYER_VULKAN_LAYER_CONTROLLER_H_
+
+#include "OrbitBase/Logging.h"
+#include "VulkanLayerProducerImpl.h"
+#include "absl/base/casts.h"
+#include "vulkan/vk_layer.h"
+#include "vulkan/vulkan.h"
+
+namespace orbit_vulkan_layer {
+
+/**
+ * This class controls the logic of this layer. For the intercepted Vulkan functions,
+ * it provides an `On*` function (e.g. for `vkQueueSubmit` there is `OnQueueSubmit`) that delegates
+ * to the driver/next layer (see `DispatchTable`) and calls the required functions for this layer to
+ * function properly. So it ties together the classes like the `SubmissionTracker` or the
+ * `TimerQueryPool`.
+ * In particular it executes the bootstrapping code (OnCreateInstance/Device) and the enumerations
+ * required by every Vulkan layer, to describe the layer as well as the extensions it uses.
+ *
+ * Usage: For an intercepted Vulkan function "X" in the layers entry (Main.cpp), `OnX` needs to be
+ * called on this controller.
+ *
+ * Note the main reason not to expose the Vulkan functions directly in this class is that this
+ * allows us to write tests. Those tests can check if we glue the code correctly together
+ * and if we do the proper bootstrapping.
+ */
+template <class DispatchTable, class QueueManager, class DeviceManager, class TimerQueryPool,
+          class SubmissionTracker>
+class VulkanLayerController {
+ public:
+  // Layer metadata. This must be in sync with the json file in the resources.
+  static constexpr const char* kLayerName = "ORBIT_VK_LAYER";
+  static constexpr const char* kLayerDescription = "Provides GPU insights for the Orbit Profiler";
+
+  static constexpr uint32_t kLayerImplVersion = 1;
```

Comment:
Not clear if this is part of the layer metadata? Remove the empty line above?

Reply:
Done.